---
title: "20251202 Repeat gene tree analysis with bootstrap values"
author: "Bin He"
date: "2025-12-02 updated `r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(ggtree)
require(treeio)
require(seqinr)
require(tidyverse)
```

```{r}
# set plotting theme to theme_bw(base_size = 16) and save the current settings
# to `old`
old <- theme_set(theme_bw(base_size = 16))
```

## Goal
Using a random selection of representative proteins from each of the PhoU and PitR homoog group, I performed phylogenetic reconstruction with boostrapping using RAxML. The computation was performed on Galaxy (history [link](https://usegalaxy.org/u/binhe-lab/h/phou-gene-tree)


1. We have used `CD-HIT` to cluster both sets at 70% identity, which reduced each set to 1/4 of the original size. But this still left us ~8000 and ~5000 sequences in each set, respectively.
2. Using the above clustered sets as the starting point, we will next select proteins based on species.

## Data
Read in the protein sequences after clustering from the `../input/fasta-sequences` folder using `seqinr` and extract the sequence names.
```{r}
# read in the protein sequences after clustering
fasta_files <- list.files(path = "../data/fasta-sequences", 
                          pattern = "*-clustered.fa$", full.names = TRUE)
names(fasta_files) <- c("PitR", "PhoU", "NptA")
print(fasta_files) # double check that the names and files match
clustered_seqs <- map(fasta_files, \(f) read.fasta(f, seqtype = "AA"))
# extract the sequence names  
seq_names <- map(clustered_seqs, \(x) getName(x)) |>
  # split the names by | into three columns and bind the three lists as rows
  lapply(\(x) tibble(name = str_split(x, "\\|", simplify = TRUE)[,1])) |>
  list_rbind(names_to = "group")
```

Read in sequence information
```{r}
info_files <- list.files(path = "../data/fasta-sequences", 
                          pattern = "*-table.tsv.gz$", full.names = TRUE)
names(info_files) <- c("PitR", "PhoU", "NptA")
print(info_files) # double check that the names and files match
seq_info <- map(info_files, \(f) read_tsv(f, show_col_types = FALSE)) |>
  list_rbind(names_to = "group")
# filter the sequences to only include those in the clustered set
clustered_seq_info <- seq_info |>
  filter(Accession %in% seq_names$name) |>
  # split the `Tax Name` column into three, with the first column containing
  # the genus and the second the species, and the remaining going to the 
  # third column (typically strain or subspecies and other info)
  separate_wider_delim(`Tax Name`, delim = " ",
                        names = c("Genus", "Species", "Other"), 
                        too_few = "align_start", too_many = "merge")
```

Because we want to make sure that the _S. aureus_ sequences are included, we will keep them in a separate table and then merge them back in later.
```{r}
# identify the _S. aureus_ sequences
staph_aureus_unip_ids = c("PitR" = "A0A0H3K7D8", 
                          "PhoU" = "A0A0H3K808",
                          "NptA" = "A0A0H3K9X3")
```

### For gene tree
Note that the original sampling was done in `20250417-select-representative-protein-and-gene-tree-plot.Rmd`, where I randomly selected 100 and 200 proteins from the PitR and PhoU sets, respectively. I also included the _S. aureus_ sequences in the list. Here, I'm importing the sequence information for the purpose of plotting the gene tree later.
```{r}
seq_for_tree <- read_tsv(file = "../data/fasta-sequences/20250425-select-proteins-for-tree.tsv")
seq_for_tree_2 <- read_tsv(file = "../data/fasta-sequences/20250425-select-proteins-for-tree-2.tsv")
seq_for_tree_3 <- read_tsv(file = "../data/fasta-sequences/20250425-select-proteins-for-tree-3.tsv")
```


## Plot gene tree
> The above selected protein sequences were aligned using MAFFT V7.526 with the E-INS-i algorithm optimized for divergent sequences with multiple conserved domains. I also tried MSAprobs and the result looked superficially similar.
> Gene tree reconstruction was performed using RAXML v8.2.12 on the Galaxy server. I used "LG+G" as the protein substitution model and matrix. I chose the `-f` option to run a rapid bootstrap, but the result is not in the output. If we want the bootstrap values, I can get it by running the program locally.

In this analysis, I repeated the above analysis but changing the parameter of the RAxML run to include a bootstrap analysis. Details of the RAxML run are as follows:

```bash
raxmlHPC-PTHREADS -T 16 -s /jetstream2/scratch/main/jobs/72980314/inputs/dataset_8c8cb86e-5e9e-4579-aeb1-285e82431c26.dat -n galaxy -m PROTGAMMALG -p 12345 -N 200 -B 0.03 -c 25 -f a -K GTR -W 100 -x 12345
```

Important parameters:
- `-N 200`: perform 200 bootstrap replicates (500 replicates for set 1, which is shown in the main figure)
- `-f a`: rapid bootstrap analysis and search for best-scoring ML tree in one
- `-x 12345`: random seed for bootstrapping

Essentially, the combination of `-f a` and `-x 12345` and `-N 200` instruct the program to perform bootstrapping. In my previous analysis, I used the default parameters in the Galaxy instance of RAxML, which has `-f D` and `-N 1`, meaning no bootstrapping was performed.

### LG+G model, set 1
```{r}
gene_tree_LG1 <- read.newick(
  "../output/gene-tree/20251202-PhoU-PitR-select-raxml-PROTGAMMA-LG-best-ML-tree-1.nwk",
  node.label = "support") %>% 
  as_tibble() %>% 
  #mutate(label = gsub("_", " ", label)) %>% 
  left_join(seq_for_tree, by = c("label" = "Accession")) %>% 
  as.treedata()
```

plot the tree, with labels for the _S. aureus_ sequences
```{r}
ggtree(gene_tree_LG1, layout = "equal_angle") +
  geom_tiplab(aes(subset = label %in% c("A0A0H3K808", "A0A0H3K7D8"), 
                  angle = angle, label = paste("S. aureus", group, sep = " ")), 
              size = 3.5, geom = "label", label.size = 0, hjust = -0.1) +
  #geom_treescale(x = 0, width = 0.2, linesize = 1.2) +
  geom_tippoint(aes(color = group), size = 1.5) +
  geom_label(aes(label = ifelse(support >= 85, support, NA)), 
            hjust = -0.4, color = "gray10", size = 3, 
            fill = rgb(0,0,0,0.1), label.size = 0) +
  scale_color_manual("Group", values =  c("PhoU" = "#6a5acd",
                                 "PitR" = "#d14949")) +
  #guides(color = guide_legend(byrow = TRUE)) +
  theme(legend.position = c(0.12, 0.13),
        legend.title = element_text(size = rel(1.2)),
        legend.text = element_text(size = rel(1.2)))
ggsave("../output/gene-tree/20251202-PhoU-PitR-select-raxml-PROTGAMMA-LG-best-ML-tree.png", 
       width = 8, height = 8)
```

### LG+G model, set 2
Read tree and add sequence information
```{r}
gene_tree_LG2 <- read.newick(
  "../output/gene-tree/20251202-PhoU-PitR-select-raxml-PROTGAMMA-LG-best-ML-tree-2.nwk",
  node.label = "support") %>% 
  as_tibble() %>% 
  #mutate(label = gsub("_", " ", label)) %>% 
  left_join(seq_for_tree_2, by = c("label" = "Accession")) %>% 
  as.treedata()
```

plot the tree, with labels for the _S. aureus_ sequences
```{r}
ggtree(gene_tree_LG2, layout = "equal_angle") +
  #geom_treescale(x = 0, width = 0.2, linesize = 1.2) +
  geom_tippoint(aes(color = group), size = 1.5) +
  geom_label(aes(label = ifelse(support >= 85, support, NA)), 
            hjust = -0.4, color = "gray10", size = 3, 
            fill = rgb(0,0,0,0.1), label.size = 0) +
  scale_color_manual("Group", values =  c("PhoU" = "#6a5acd",
                                 "PitR" = "#d14949")) +
  geom_tiplab(aes(subset = label %in% c("A0A0H3K808", "A0A0H3K7D8"), 
                  angle = angle, label = paste("S. aureus", group, sep = " ")), 
              size = 3.5, geom = "label", label.size = 0, hjust = -0.1) +
  #guides(color = guide_legend(byrow = TRUE)) +
  theme(legend.position = c(0.12, 0.13),
        legend.title = element_text(size = rel(1.2)),
        legend.text = element_text(size = rel(1.2)))
ggsave("../output/gene-tree/20251202-PhoU-PitR-select-raxml-PROTGAMMA-LG-best-ML-tree-2.png", 
       width = 8, height = 8)
```

### LG+G model, set 3
Read tree and add sequence information
```{r}
gene_tree_LG3 <- read.newick(
  "../output/gene-tree/20251202-PhoU-PitR-select-raxml-PROTGAMMA-LG-best-ML-tree-3.nwk",
  node.label = "support") %>% 
  as_tibble() %>% 
  #mutate(label = gsub("_", " ", label)) %>% 
  left_join(seq_for_tree_3, by = c("label" = "Accession")) %>% 
  as.treedata()
```

plot the tree, with labels for the _S. aureus_ sequences
```{r}
ggtree(gene_tree_LG3, layout = "equal_angle") +
  geom_tiplab(aes(subset = label %in% c("A0A0H3K808", "A0A0H3K7D8"), 
                  angle = angle, label = paste("S. aureus", group, sep = " ")), 
              size = 3.5, geom = "label", label.size = 0) +
  geom_tippoint(aes(color = group), size = 1.5) +
  geom_label(aes(label = ifelse(support >= 85, support, NA)), 
            hjust = -0.4, color = "gray10", size = 3, 
            fill = rgb(0,0,0,0.1), label.size = 0) +
  scale_color_manual("Group", values =  c("PhoU" = "#6a5acd",
                                 "PitR" = "#d14949")) +
  #guides(color = guide_legend(byrow = TRUE)) +
  theme(legend.position = c(0.12, 0.13),
        legend.title = element_text(size = rel(1.2)),
        legend.text = element_text(size = rel(1.2)))
ggsave("../output/gene-tree/20251202-PhoU-PitR-select-raxml-PROTGAMMA-LG-best-ML-tree-3.png", 
       width = 8, height = 8)
```
