---
title: "20250407-select-representative-proteins"
author: "Bin He"
date: "2025-04-07 updated `r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(seqinr)
require(tidyverse)
```

```{r}
# set plotting theme to theme_bw(base_size = 16) and save the current settings
# to `old`
old <- theme_set(theme_bw(base_size = 16))
```

## Goal
Select 100-300 representative proteins from each of the PhoU and PitR related proteins. We will use two criteria to select:
1. We have used `CD-HIT` to cluster both sets at 70% identity, which reduced each set to 1/4 of the original size. But this still left us ~8000 and ~5000 sequences in each set, respectively.
2. Using the above clustered sets as the starting point, we will next select proteins based on species.

## Data
Read in the protein sequences after clustering from the `../input/fasta-sequences` folder using `seqinr` and extract the sequence names.
```{r}
# read in the protein sequences after clustering
fasta_files <- list.files(path = "../data/fasta-sequences", 
                          pattern = "*-clustered.fa$", full.names = TRUE)
names(fasta_files) <- c("PitR", "PhoU", "NptA")
print(fasta_files) # double check that the names and files match
clustered_seqs <- map(fasta_files, \(f) read.fasta(f, seqtype = "AA"))
# extract the sequence names  
seq_names <- map(clustered_seqs, \(x) getName(x)) |>
  # split the names by | into three columns and bind the three lists as rows
  lapply(\(x) tibble(name = str_split(x, "\\|", simplify = TRUE)[,1])) |>
  list_rbind(names_to = "group")
```

Read in sequence information
```{r}
info_files <- list.files(path = "../data/fasta-sequences", 
                          pattern = "*-table.tsv.gz$", full.names = TRUE)
names(info_files) <- c("PitR", "PhoU", "NptA")
print(info_files) # double check that the names and files match
seq_info <- map(info_files, \(f) read_tsv(f, show_col_types = FALSE)) |>
  list_rbind(names_to = "group")
# filter the sequences to only include those in the clustered set
clustered_seq_info <- seq_info |>
  filter(Accession %in% seq_names$name) |>
  # split the `Tax Name` column into three, with the first column containing
  # the genus and the second the species, and the remaining going to the 
  # third column (typically strain or subspecies and other info)
  separate_wider_delim(`Tax Name`, delim = " ",
                        names = c("Genus", "Species", "Other"), 
                        too_few = "align_start", too_many = "merge")
```

Because we want to make sure that the _S. aureus_ sequences are included, we will keep them in a separate table and then merge them back in later.
```{r}
# identify the _S. aureus_ sequences
staph_aureus_unip_ids = c("PitR" = "A0A0H3K7D8", 
                          "PhoU" = "A0A0H3K808",
                          "NptA" = "A0A0H3K9X3")
```


## Select representative proteins
Group by genus and select the first sequence in each group
```{r}
genus_representative <- clustered_seq_info |>
  distinct(group, Genus, .keep_all = TRUE)
count(genus_representative, group)
```

After both `CD-HIT` clustering and species filtering, we are left with 1k and 2k sequences in the PitR and PhoU sets. 

Now we will check if the _S. aureus_ representatives are in the list.
```{r}
# are they all in the list?
all(staph_aureus_unip_ids %in% genus_representative$Accession)
```

Good. Now, we will generate a filtered fasta file containing the genus representative sequences for the PhoU and PitR homologs only. When exporting, we will rename the sequences to the following format:
">Accession PhoU|PitR Genus_species CD-HIT_cluster_id"

First, we identify the selected sequence accessions and filter the sequence list.
```{r}
# merge the PitR and PhoU related sequences, leaving out NptA
grouped_seqs <- do.call(c, clustered_seqs[c("PitR", "PhoU")])

# Filter sequences: get only those whose name matches selected accessions
grouped_seqs_acc <- getName(grouped_seqs) |> str_split("\\|", simplify = TRUE)
grouped_seqs_acc <- grouped_seqs_acc[,1]
select_seqs_acc <- genus_representative |> filter(group != "NptA") |> pull(Accession)
selected_seqs <- grouped_seqs[which(grouped_seqs_acc %in% select_seqs_acc)]
sprintf("Selected %d sequences from %d total", length(selected_seqs), length(grouped_seqs))
```
Next, we rename the sequences
```{r}
# Convert list to a named character vector for renaming
new_names <- genus_representative |>
  filter(group != "NptA") |>
  mutate(new_name = paste(Accession, group, Genus, Species, sep = " ")) |>
  select(Accession, new_name) |>
  deframe()

# get the names of the selected sequences
selected_seqs_acc <- getName(selected_seqs) |> str_split("\\|", simplify = TRUE)
selected_seqs_acc <- selected_seqs_acc[,1]

# check to make sure that the selected sequences are in the new names
all(selected_seqs_acc %in% names(new_names))

# export the selected sequences to a fasta file
write.fasta(sequences = selected_seqs, 
            names = new_names[selected_seqs_acc], 
            file.out = "../data/fasta-sequences/20250417-representative-proteins.fa")
```

